### PR TITLE
Add docs for the WebLink component

### LINKED ISSUE #7515: Add docs for the new WebLink component
see symfony/symfony#21478

### PR DESCRIPTION
**Pull request by dunglas**
The WebLink component is available since Symfony 3.3, but I never took the time to add the docs (however, a blog post explaining how to use it was available).

This documentation is based on https://dunglas.fr/2017/10/symfony-4-http2-push-and-preloading/.
If necessary, I can grant any copyright regarding this post to the Symfony project.

symfony/symfony#21478
https://github.com/symfony/symfony/pull/22273
Closes #7515.

### CODE REVIEW

**Comment by dunglas** [comment]
Do you know how to fix this RST error? https://travis-ci.org/symfony/symfony-docs/builds/426930677

**Comment by javiereguiluz** [comment]
About the RST issue, you must use this alternative format for the link of the line 64:

```rst
`Bootstrap`_
```

**Comment by xabbuh** [comment]
looks like the failure was related to some whitespace issues (see 38fda884205e441463aa424e9927a6fabf621b05)

**Comment by dunglas** [comment]
Thanks a lot @xabbuh!

File: weblink.rst [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+Symfony natively supports `Web Linking`_. It is especially useful to improve
+the performance of your application by leveraging the HTTP/2 protocol and
+preloading capabilities of modern web browsers.
+
+By implementing cutting edge web standards, namely `HTTP/2 Server Push`_ and
+W3C's `Resource Hints`_, the WebLink component
+brings great opportunities to boost webapp's performance.
+
+Thanks to WebLink, HTTP/2 (**h2**) servers are able to push resources to clients
+before they even know that they need them (think to CSS or JavaScript
```
Comment (nicolas-grekas) [inline weblink.rst]: think about

File: weblink.rst [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+files, or relations of an API resource). WebLink also enables other very
+efficient optimisations that work with HTTP 1:
```
Comment (nicolas-grekas) [inline weblink.rst]: HTTP 1.x or 1.1?

File: weblink.rst [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+-  telling the browser to fetch or to render another webpage in the
+   background ;
+-  init early DNS lookups, TCP handshakes or TLS negotiations
```
Comment (nicolas-grekas) [inline weblink.rst]: initializing

File: weblink.rst [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+Let's discover how easy it is to use it and the real life benefits you
```
Comment (nicolas-grekas) [inline weblink.rst]: I'd suggest removing this sentence

File: weblink.rst [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+can expect.
+
+To benefit from HTTP/2 Server Pushes, a HTTP/2 server and a HTTPS connection
```
Comment (nicolas-grekas) [inline weblink.rst]: s/a/an

File: weblink.rst [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+are mandatory (even in local).
```
Comment (nicolas-grekas) [inline weblink.rst]: s/mandatory/required + (even local ones)

File: weblink.rst [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+Both Apache, Nginx and Caddy support these protocols.
+Be sure they are properly configured before reading.
+
+Alternatively, you can use the `Docker installer and runtime for
+Symfony`_ provided by Kévin Dunglas (community supported).
+
+It includes everything you need to run Symfony
```
Comment (nicolas-grekas) [inline weblink.rst]: not sure about this part + the below instruction: looks redundant with the readme of symfony-docker

File: weblink.rst [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+
+Symfony is smart enough to download Twig, to automatically register it,
+and to enable Symfony features requiring the library.
+It also generates a base HTML5 layout in the ``templates/`` directory.
+
+Now, download Bootstrap_, extract the archive and copy the file
+``dist/css/bootstrap.min.css`` in the ``public/`` directory of our
+project.
+
+Symfony comes with a `nice integration with of the most popular CSS framework`_.
```
Comment (nicolas-grekas) [inline weblink.rst]: s/with//

File: weblink.rst [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+
+    In a real project, you should use Yarn or NPM with
+    :doc:`Symfony Encore </frontend/encore/bootstrap>`
+    to install Bootstrap.
+
+Now, it's time to create the template of our homepage:
+
+.. code-block:: html
+
+   <!DOCTYPE html>
```
Comment (nicolas-grekas) [inline weblink.rst]: missing path where to put the file?

File: weblink.rst [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+.. code-block:: yaml
+
+   # config/routes.yaml
+   index:
+       path: /
+       defaults:
+         _controller: 'Symfony\Bundle\FrameworkBundle\Controller\TemplateController::templateAction'
+         template: 'homepage.html.twig'
+
+Refresh your browser, this nice homepage should appear:
```
Comment (nicolas-grekas) [inline weblink.rst]: s/nice//

File: weblink.rst [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+
+.. image:: /_images/components/weblink/http2-server-push.png
+
+As you can see (Initiator: Push), both
+responses have been sent directly by the server.
+``bootstrap.min.css`` has started to be received before the browser even requested it!
+
+.. note::
+
+    Google Chrome provides a nice interface to debug HTTP/2 connections.
```
Comment (nicolas-grekas) [inline weblink.rst]: s/nice//

File: weblink.rst [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+    Open ``chrome://net-internals/#http2`` to start the tool.
+
+How does it works?
+~~~~~~~~~~~~~~~~~~
+
+The WebLink component tracks Link HTTP headers to add to the response.
```
Comment (nicolas-grekas) [inline weblink.rst]: missing backticks around Link

File: weblink.rst [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+When using the ``preload()`` helper, a ``Link`` header
+with a `preload`_
+``rel`` attribute is added to the response:
+
+.. image:: /_images/components/weblink/response-headers.png
+
+According to `the Preload specification`_,
+when a HTTP/2 server detects that the original (HTTP 1) response
```
Comment (nicolas-grekas) [inline weblink.rst]: an HTTP + HTTP 1.x

File: weblink.rst [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+   {# ... #}
+       <link rel="stylesheet" href="{{ preload('/bootstrap.min.css', {nopush: true}) }}">
+   {# ... #}
+
+Before using HTTP/2 Push, be sure to read `this great article`_ about
+known issues, cache implications and the state of the support in popular
+browsers.
+
+In addition to HTTP/2 Push and preloading, the WebLink component also
+provide some helpers to send `Resource
```
Comment (nicolas-grekas) [inline weblink.rst]: provides

File: weblink.rst (hunk 11 above)
Comment (OskarStark) [reply weblink.rst]: Afterwards s/a nice interface/an interface/

**Comment by OskarStark** [comment]
Hey @dunglas,

you mention some twig helpers in your blog post, but they are not included in the component itself, is a bridge or anything else available? Or am I missing sth?

**Comment by dunglas** [comment]
Yes, they belong in the TwigBridge.

File: weblink.rst (hunk 7 above)
Comment (dunglas) [reply weblink.rst]: removed

**Comment by dunglas** [comment]
Than you very much @nicolas-grekas, changes applied.

**Comment by javiereguiluz** [comment]
Thanks Kévin for contributing this doc (and the component too!). I made some changes while merging, specially on the main `weblink.rst` article. Instead of a self-contained tutorial that installs Bootstrap, creates a new Symfony project, etc. I streamlined to assume that you already have a working Symfony app so we can focus the explanation on the preloading and resource hints features. Thanks!

**Pull request merged**